src/node.rs

```diff
@@ -36,46 +65,91 @@ impl NetworkManager {
         NetworkManager { socket, config }
     }
 
+    // pub async fn send_message(
+    //     &self,
+    //     message: &Message,
+    //     dst: SocketAddr,
+    // ) -> Result<(), KademliaError> {
+    //     debug!("Sending message to {}: {:?}", dst, message);
+    //     let serialized = bincode::serialize(message)?;
+    //     self.socket.send_to(&serialized, dst).await.map_err(|e| {
+    //         error!("Failed to send message to {}: {:?}", dst, e);
+    //         KademliaError::Network(e)
+    //     });
+    //     Ok(())
+    // }
+
     pub async fn send_message(
         &self,
         message: &Message,
         dst: SocketAddr,
     ) -> Result<(), KademliaError> {
         debug!("Sending message to {}: {:?}", dst, message);
-        let serialized = bincode::serialize(message)?;
-        self.socket.send_to(&serialized, dst).await?;
-        Ok(())
+
+        exponential_backoff(
+            || {
+                let message = message.clone();
+                let socket = Arc::clone(&self.socket);
+                Box::pin(async move {
+                    let serialized = bincode::serialize(&message)?;
+                    socket
+                        .send_to(&serialized, dst)
+                        .await
+                        .map(|_| ())
+                        .map_err(|e| {
+                            error!("Failed to send message to {}: {:?}", dst, e);
+                            KademliaError::Network(e)
+                        })
+                }) as Pin<Box<dyn Future<Output = Result<(), KademliaError>>>>
+            },
+            3, // Number of retries
+        )
+        .await
     }
 
     pub async fn receive_message(&self) -> Result<(Message, SocketAddr), KademliaError> {
-        let mut buf = vec![0u8; 1024];
-        let (size, src) = self.socket.recv_from(&mut buf).await?;
-        let message: Message = bincode::deserialize(&buf[..size])?;
-        debug!("Received message from {}: {:?}", src, message);
-        Ok((message, src))
+        let socket = Arc::clone(&self.socket);
+        exponential_backoff(
+            move || {
+                Box::pin(async move {
+                    let mut buf = vec![0u8; 1024];
+                    let (size, src) = socket.recv_from(&mut buf).await.map_err(|e| {
+                        error!("Failed to receive message: {:?}", e);
+                        KademliaError::Network(e)
+                    })?;
+                    let message: Message = bincode::deserialize(&buf[..size])?;
+                    debug!("Received message from {}: {:?}", src, message);
+                    Ok((message, src))
+                })
+                    as Pin<Box<dyn Future<Output = Result<(Message, SocketAddr), KademliaError>>>>
+            },
+            3, // Number of retries
+        )
+        .await
     }
 }
 
 // Storage Manager
 pub struct StorageManager {
```

Comment:
This should be a trait so we can have in-memory storage or other persistent storage implementations to swap out.